### PR TITLE
25 enable clear var in terminal

### PR DESCRIPTION
**Pull request by MarkRamp**
### SUMMARY
Implements necessary changes for #[ISSUE-ID]

*Reviewers:* @numere-org/maintainers

### IMPLEMENTATION
* Implementation: (*Describe, what you've changed*) 
* Implementation test: (*Describe the type of test, which you performed, and if it was successful*)

### DOCUMENTATION
* [x] ChangesLog updated
* [x] Code changes commented
* **Documentation articles:**
    * [ ] corresponding documentation articles updated
    * [ ] new documentation articles created
    * [x] not needed
* **Language files:**
    * [ ] corresponding language files updated
    * [x] not needed

----------------

### TESTS BY REVIEWERS
* [ ] Added to the automatic SW tests
* [x] Tested manually

### CODE REVIEW

**Comment by durveshps** [comment]
Ich habe die Grund-Funktionalität umgesetzt.

File: kernel/core/strings/stringvarfactory.cpp [hunk 1]
```diff
@@ -746,5 +746,33 @@ namespace NumeRe
             m_mStringVars.erase(iter);
     }
 
+
+    /////////////////////////////////////////////////
+    /// \brief This public member function removes all
+    /// string variables which don't start with ~,_
+    ///
+    /// \param sVar const string&
```
Comment (durveshps) [inline kernel/core/strings/stringvarfactory.cpp:754]: should be corrected

File: kernel/core/commandfunctions.hpp [hunk 2]
```diff
@@ -3142,6 +3142,33 @@ static CommandReturnValues cmd_clear(string& sCmd)
...
+
+        static vector<string> sLegacyVar = {"t", "x", "y", "z", "ncols", "nlens", "nlines", "nrows", "ans"};
+        Parser& _parser = NumeReKernel::getInstance()->getParser();
+        varmap_type mVariables = _parser.GetVar();
+
+        // check if user-defined variable is legacy variable
+        for (auto iter = mVariables.begin(); iter != mVariables.end(); ++iter) {
+            bool isLegacyVariable = false;
+            for (int i = 0; i != sLegacyVar.size(); i++) {
+                // if ( sLegacyVar[i] == iter->first || iter->first[0] == "~" || iter->first[0] == "_")
```
Comment (durveshps) [inline kernel/core/commandfunctions.hpp:3158]: can be removed

File: kernel/core/strings/stringvarfactory.cpp [hunk 3]
```diff
@@ -746,5 +746,33 @@ namespace NumeRe
...
+	{
+	    NumeReKernel::getInstance()->print(iter->first);
+	    if (iter->first.substr(0, 1) == "_" || iter->first.substr(0, 1) == "~")
+	    {
+		continue;
+	    }
+	    else
+	    {
+		m_mStringVars.erase(iter);
+	    }
```
Comment (numeredev) [inline kernel/core/strings/stringvarfactory.cpp:773]:
```suggestion
        if (!iter->first.starts_with("_~"))
	        m_mStringVars.erase(iter);
```

This is, what you probably meant.

File: kernel/core/strings/stringvarfactory.cpp (hunk 1 above)
Comment (numeredev) [reply kernel/core/strings/stringvarfactory.cpp:754]: Yes please

File: kernel/core/commandfunctions.hpp [hunk 4]
```diff
@@ -3142,6 +3142,33 @@ static CommandReturnValues cmd_clear(string& sCmd)
 
         return cmd_delete(sCommand);
     }
+    else if (cmdParser.hasParam("var"))
+    {
+        NumeRe::StringParser& _stringParser = NumeReKernel::getInstance()->getStringParser();
+        _stringParser.clearStringVar();
+
+        static vector<string> sLegacyVar = {"t", "x", "y", "z", "ncols", "nlens", "nlines", "nrows", "ans"};
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3150]:
```suggestion
        static std::vector<std::string> dafaultVars = {"t", "x", "y", "z", "ncols", "nlens", "nlines", "nrows", "ans"};
```

File: kernel/core/commandfunctions.hpp [hunk 5]
```diff
@@ -3142,6 +3142,33 @@ static CommandReturnValues cmd_clear(string& sCmd)
 
         return cmd_delete(sCommand);
     }
+    else if (cmdParser.hasParam("var"))
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3145]:
```suggestion
    else if (cmdParser.hasParam("var") && NumeReKernel::getInstance()->getDebugger().getStackSize() == 0)
```
You want to ensure that this is only allowed within the terminal out of any procedures or flow control statements.

**Review by numeredev — CHANGES_REQUESTED** [review]
Please check, whether those suggestions work. Feel free to ask questions 😉

**Comment by MarkRamp** [comment]
> Please check, whether those suggestions work. Feel free to ask questions 😉

I have made all of the changes, clear -memory will now also call -var, becuase memory word is superior to variable, I think its ready to be merged.

File: kernel/core/strings/stringvarfactory.cpp [hunk 6]
```diff
@@ -715,5 +715,24 @@ namespace NumeRe
...
+    void StringVarFactory::clearStringVar()
+    {
+	if (!m_mStringVars.size())
+            return;
+
+	for (auto iter = m_mStringVars.begin(); iter != m_mStringVars.end(); ++iter)
+	{
+	    if (!iter->first.starts_with("_~"))
+	        m_mStringVars.erase(iter);
+	}
```
Comment (numeredev) [inline kernel/core/strings/stringvarfactory.cpp:735]:
```suggestion
	    if (!m_mStringVars.size())
            return;

	    for (auto iter = m_mStringVars.begin(); iter != m_mStringVars.end(); ++iter)
	    {
	        if (!iter->first.starts_with("_~"))
	            m_mStringVars.erase(iter);
	    }
```

File: kernel/core/commandfunctions.hpp [hunk 7]
```diff
@@ -3115,6 +3146,12 @@ static CommandReturnValues cmd_clear(string& sCmd)
         // Clear also the string table
         _data.clearStringElements();
 
+	// Clear also user-defined variables
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3149]:
```suggestion
	// Clear also user-defined variables if called from terminal
```

File: kernel/core/commandfunctions.hpp [hunk 8]
```diff
@@ -3104,6 +3131,10 @@ static CommandReturnValues cmd_clear(string& sCmd)
 
         return cmd_delete(sCommand);
     }
+    else if (cmdParser.hasParam("var") && NumeReKernel::getInstance()->getDebugger().getStackSize() == 0)
+    {
+	clear_variables();
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3136]:
```suggestion
	    clear_variables();
```

File: kernel/core/commandfunctions.hpp [hunk 9]
```diff
@@ -3077,6 +3077,33 @@ static CommandReturnValues cmd_delete(string& sCmd)
...
+static void clear_variables()
+{
+    NumeRe::StringParser& _stringParser = NumeReKernel::getInstance()->getStringParser();
+    _stringParser.clearStringVar();
+
+    static std::vector<std::string> defaultVars = {"t", "x", "y", "z", "ncols", "nlens", "nlines", "nrows", "ans"};
+    Parser& _parser = NumeReKernel::getInstance()->getParser();
+    varmap_type mVariables = _parser.GetVar();
+
+    // check if user-defined variable is legacy variable and call is from terminal
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3096]: Comment does not make any sense. It should be something like `remove variables, which are not default variables and no temporary variables`

File: kernel/core/commandfunctions.hpp [hunk 10]
```diff
@@ -3077,6 +3077,33 @@ static CommandReturnValues cmd_delete(string& sCmd)
...
+    for (auto iter = mVariables.begin(); iter != mVariables.end(); ++iter)
+    {
+	bool isLegacyVariable = iter->first.starts_with("_~");
+	bool isCallingFromTerminal = std::find(defaultVars.begin(), defaultVars.end(), iter->first) == defaultVars.end();
+	if (!isLegacyVariable && isCallingFromTerminal)
+	    _parser.RemoveVar(iter->first);
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3102]:
```suggestion
	    bool isTemporaryVar = iter->first.starts_with("_~");
	    bool isDefaultVar = std::find(defaultVars.begin(), defaultVars.end(), iter->first) != defaultVars.end();
	    if (!isTemporaryVar && !isDefaultVar)
	        _parser.RemoveVar(iter->first);
```

**Review by numeredev — CHANGES_REQUESTED** [review]
Some minor changes still open. There should also be an update to `data_and_fileops.nhlp` in both languages to explain the behavior of `clear -var`

File: kernel/core/strings/stringvarfactory.cpp (hunk 6 above)
Comment (numeredev) [reply kernel/core/strings/stringvarfactory.cpp:735]: Addendum: obviously this file contains a mixture of whitespaces and tabs for some reason. Do only apply indentation suggestions, if they are reasonable

**Comment by MarkRamp** [comment]
Translation PR 
https://github.com/numere-org/Translation/pull/20

File: kernel/core/commandfunctions.hpp [hunk 11]
```diff
@@ -3104,6 +3131,10 @@ static CommandReturnValues cmd_clear(string& sCmd)
 
         return cmd_delete(sCommand);
     }
+    else if (cmdParser.hasParam("var") && NumeReKernel::getInstance()->getDebugger().getStackSize() == 0)
+    {
+	    clear_variables();
+    }
     else if (cmdParser.hasParam("memory"))
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3138]:
```suggestion
    else if (cmdParser.hasParam("memory") && NumeReKernel::getInstance()->getDebugger().getStackSize() == 0)
```

File: kernel/core/commandfunctions.hpp [hunk 12]
```diff
@@ -3115,6 +3146,12 @@ static CommandReturnValues cmd_clear(string& sCmd)
         // Clear also the string table
         _data.clearStringElements();
 
+	// Clear also user-defined variables if called from terminal
+	if (NumeReKernel::getInstance()->getDebugger().getStackSize() == 0)
+	{
+	    clear_variables();
+	}
```
Comment (numeredev) [inline kernel/core/commandfunctions.hpp:3153]:
```suggestion
	clear_variables();
```

**Pull request merged**